styx-service-common/src/main/java/com/spotify/styx/storage/DatastoreStorage.java

```diff
@@ -969,32 +934,67 @@ static StringValue jsonValue(Object o) throws JsonProcessingException {
       throws IOException, E {
-    final StorageTransaction tx = newTransaction();
-    try {
-      final T value = f.apply(tx);
-      tx.commit();
-      return value;
-    } catch (DatastoreIOException e) {
-      throw new TransactionException(e.getCause());
-    } catch (DatastoreException e) {
-      throw new TransactionException(e);
-    } finally {
-      if (tx.isActive()) {
-        tx.rollback();
+
+    var retrySettings = datastore.getOptions().getRetrySettings();
+    var backoff = new ExponentialBackOff.Builder()
+        .setInitialIntervalMillis((int) retrySettings.getInitialRetryDelay().toMillis())
+        .setMaxIntervalMillis((int) retrySettings.getMaxRetryDelay().toMillis())
+        .setMultiplier(retrySettings.getRetryDelayMultiplier())
+        .setMaxElapsedTimeMillis((int) retrySettings.getMaxRetryDelay().toMillis())
+        .build();
+
+    for (int attempt = 1; ; attempt++) {
+      try {
+        return runInTransaction(f);
+      } catch (DatastoreIOException e) {
+        if (!isRetryableTransactionException(e.getCause())) {
+          throw e;
+        }
+        if (retrySettings.getMaxAttempts() > 0 && attempt >= retrySettings.getMaxAttempts()) {
+          throw e;
+        }
+        // Back-off sleep before retrying
+        var sleepMillis = backoff.nextBackOffMillis();
+        if (sleepMillis == STOP) {
+          throw e;
+        }
+        LOG.debug("datastore transaction exception, attempt #{}, retrying in {}ms", attempt, sleepMillis, e);
+        try {
+          Thread.sleep(sleepMillis);
+        } catch (InterruptedException ex) {
+          Thread.currentThread().interrupt();
+        }
       }
     }
   }
 
-  private StorageTransaction newTransaction() throws TransactionException {
-    final CheckedDatastoreTransaction transaction;
+  private <T, E extends Exception> T runInTransaction(TransactionFunction<T, E> f) throws IOException, E {
+    var transaction = datastore.newTransaction();
     try {
-      transaction = datastore.newTransaction();
-    } catch (DatastoreIOException e) {
-      throw new TransactionException(e.getCause());
+      var result = f.apply(storageTransactionFactory.apply(transaction));
+      transaction.commit();
+      return result;
+    } finally {
+      if (transaction.isActive()) {
+        try {
+          transaction.rollback();
+        } catch (DatastoreIOException ex) {
+          LOG.debug("Exception on rollback", ex);
+        }
+      }
     }
-    return storageTransactionFactory.apply(transaction);
+  }
+
+  private boolean isRetryableTransactionException(DatastoreException e) {
+    return
+        // code 10 ABORTED means that the transaction was aborted due to a conflict and can be retried.
+        e.getCode() == 10
+        || "ABORTED".equals(e.getReason())
+        // The datastore client has a retry bug where it incorrectly retries ABORTED lookup operations
```

Comment:
add a todo to remind us that we should remove this check after datastore client fixes the bug.

Reply:
👍